### ISSUE TITLE
[FSDP] Training gets slower as iterations increase?

### ISSUE THREAD

**Issue by woodyx218** [facebookresearch/fairscale#1102]
Hi, I test FSDP on a vit_gigantic_patch14_224 from 'timm' library and observe a very strange behavior. I use 'tqdm' to count the time used so far and the time to finish. However, the time to finish keeps increasing and the training (denoted by it/s) is getting slower and slower as time passes.

I am using the latest fairscale (0.4.13) and here is [my code](https://drive.google.com/file/d/1SFLlZHZ2xnmtnE228Kqr40ax3a7nnqAc/view?usp=share_link). Maybe it is due to auto_wrap not being used?

**Comment by min-xu-ai** [comment]
> (1) we flatten the parameters and suffer a possibly 10 times slowdown (2) or we does not flatten the parameters and cannot keep customized attributes during backward.

You mean the other way around above? I.e. non-flatten and 10x slowdown and flatten but can't keep attributes?

There are 2 issues here:

(1) slow down with non-flatten
(2) slower and slower as more iterations are done

Both are important but I'd like to know which one is which. Do you seem GPU memory getting used more and more as training progresses? How many GPUs do you use and if it is multiple-nodes, what kind network bandwidth is available between nodes?

Also, with you code, you only FSDP wrap the net at the top level. But to fully realize the memory saving of FSDP, you need to wrap the inner modules of your model with FSDP. Have you done that?

**Comment by woodyx218** [comment]
Yes you are right. I have corrected my typo. Both issues exist. For the first issue, I have set some benchmarks for 1 epoch (using my attached code). GPU memory usage is stable during the training. I have used 8 A100 GPU in a single node. Here B means mini_batch_size that GPU sees every step.

|                           | vit_base | vit_large |
|---------------------------|----------|-----------|
| B=8, flatten_param=False  | 5'23''   | 10'10''   |
| B=8, flatten_param=True   | 41''     | 77''      |
| B=32, flatten_param=False | 56''     | 92''      |
| B=32, flatten_param=True  | 30''     | 59''      |

For the second issue, the slowing down is observed no matter flatten_parameters=True or False. E.g. with vit_base_patch16_224, B=8, no flatten, in the first few iterations the total estimated time is 2 min (5.3it/s), but eventually it slows down to 5.5min (2.4it/s).

I was following the [tutorial](https://pytorch.org/blog/introducing-pytorch-fully-sharded-data-parallel-api/) to wrap the net, could you please point me to any reference about wrapping individual modules?

**Comment by min-xu-ai** [comment]
> I was following the [tutorial](https://pytorch.org/blog/introducing-pytorch-fully-sharded-data-parallel-api/) to wrap the net, could you please point me to any reference about wrapping individual modules?

The tutorial has both auto-wrap and manual-wrap sections. Maybe you missed that or I mis-read your code above? I think you only wrapped the top level model (net) once in your code.

Thanks for the great benchmarking results. The table makes a lot of sense since flatten_param=False means a lot of communication overhead. When B is small, those communications are done more frequently with respect to wall-clock time. When B=32, the slow down is not as bad. If you can use B=32 or bigger, I'd suggest you use that.

In terms of the slow down of it/s, I am wondering how is it measured. Is it measure by using time.time() on the CPU? In that case, initial measurement can be too good because CPU runs ahead of the GPU. Maybe you want to measure it after the initial warm-up period (say 10 minutes), then measure the it/s for the time period of (10-20min) and (20-30min) to compare if there is actually a slowdown or not?

**Comment by woodyx218** [comment]
I only wrapped once in my code. Could you point to me how to wrap individual modules? The [tutorial](https://pytorch.org/blog/introducing-pytorch-fully-sharded-data-parallel-api/) seems to only support top level wrapping (both auto-wrap and manual-wrap).

In large model training, I have to use small batch size. I can do the measuring as you suggested in the next two days.

**Comment by min-xu-ai** [comment]
> I only wrapped once in my code. Could you point to me how to wrap individual modules? The [tutorial](https://pytorch.org/blog/introducing-pytorch-fully-sharded-data-parallel-api/) seems to only support top level wrapping (both auto-wrap and manual-wrap).

You can check FSDP tests for many examples of how to wrap. (inside tests/ dir)

> In large model training, I have to use small batch size. I can do the measuring as you suggested in the next two days.

Once you do nested wrapping + activation checkpointing, you will likely be able to use larger batch size per GPU due to memory saving.

**Comment by min-xu-ai** [comment]
also, when you go to large model, the overhead of flatten_param will likely change (sometimes dramatically)

**Comment by woodyx218** [comment]
Can you please re-open this issue as new evidence is provided?

**Comment by min-xu-ai** [comment]
Unfortunately, I can only help you so far. If you want to debug and fix it yourself, I am happy to answer more questions for the code.

**Comment by woodyx218** [comment]
That's a pity since flatten_parameters seems to be an important feature, but the linear slowdown is universal (even following the official [tutorial](https://pytorch.org/tutorials/intermediate/FSDP_tutorial.html), except setting flatten_parameters=False). I may add more discoveries to this thread, and hopefully it helps future readers.
